--- a/lib/Controller/SlaveController.php	2023-04-21 08:42:55.000000000 +0200
+++ b/lib/Controller/SlaveController.php	2023-05-15 11:56:58.812196333 +0200
@@ -173,8 +173,8 @@
 		$this->logger->debug('all good. creating session');
 		$this->userSession->createSessionToken($this->request, $uid, $uid, null, IToken::REMEMBER);
 
-//		$user = $this->userManager->get($uid);
-//		$this->twoFactorManager->prepareTwoFactorLogin($user, false);
+		$user = $this->userManager->get($uid);
+		$this->twoFactorManager->prepareTwoFactorLogin($user, false);
 
 		$this->slaveService->updateUserById($uid);
 		$this->logger->debug('userdata updated on lus');
