diff --git a/apps/files_external/lib/Lib/PersonalMount.php b/apps/files_external/lib/Lib/PersonalMount.php
index b22266fce180..b04874946352 100644
--- a/apps/files_external/lib/Lib/PersonalMount.php
+++ b/apps/files_external/lib/Lib/PersonalMount.php
@@ -36,8 +36,8 @@ class PersonalMount extends ExternalMountPoint implements MoveableMount {
 	/** @var UserStoragesService */
 	protected $storagesService;
 
-	/** @var int */
-	protected $numericStorageId;
+	/** @var int id of the external storage (mount) (not the numeric id of the resulting storage!) */
+	protected $numericExternalStorageId;
 
 	/**
 	 * @param UserStoragesService $storagesService
@@ -51,7 +51,7 @@ class PersonalMount extends ExternalMountPoint implements MoveableMount {
 	public function __construct(
 		UserStoragesService $storagesService,
 		StorageConfig $storageConfig,
-		$storageId,
+		$externalStorageId,
 		$storage,
 		$mountpoint,
 		$arguments = null,
@@ -61,7 +61,7 @@ public function __construct(
 	) {
 		parent::__construct($storageConfig, $storage, $mountpoint, $arguments, $loader, $mountOptions, $mountId);
 		$this->storagesService = $storagesService;
-		$this->numericStorageId = $storageId;
+		$this->numericExternalStorageId = $externalStorageId;
 	}
 
 	/**
@@ -71,7 +71,7 @@ public function __construct(
 	 * @return bool
 	 */
 	public function moveMount($target) {
-		$storage = $this->storagesService->getStorage($this->numericStorageId);
+		$storage = $this->storagesService->getStorage($this->numericExternalStorageId);
 		// remove "/$user/files" prefix
 		$targetParts = explode('/', trim($target, '/'), 3);
 		$storage->setMountPoint($targetParts[2]);
@@ -86,7 +86,7 @@ public function moveMount($target) {
 	 * @return bool
 	 */
 	public function removeMount() {
-		$this->storagesService->removeStorage($this->numericStorageId);
+		$this->storagesService->removeStorage($this->numericExternalStorageId);
 		return true;
 	}
 }
