diff --git a/lib/Master.php b/lib/Master.php
index 1c410f8..c5a1d0f 100644
--- a/lib/Master.php
+++ b/lib/Master.php
@@ -30,10 +30,12 @@
 use OCA\GlobalSiteSelector\AppInfo\Application;
 use OCA\GlobalSiteSelector\UserDiscoveryModules\IUserDiscoveryModule;
 use OCP\Authentication\IApacheBackend;
+use OC\Authentication\TwoFactorAuth\Manager;
 use OCP\HintException;
 use OCP\Http\Client\IClientService;
 use OCP\IConfig;
 use OCP\IRequest;
+use OCP\IUserManager;
 use OCP\Security\ICrypto;
 use OCP\Server;
 use Psr\Container\ContainerExceptionInterface;
@@ -51,27 +53,33 @@
 	private GlobalSiteSelector $gss;
 	private ICrypto $crypto;
 	private Lookup $lookup;
-	private IRequest $request;
+  private IRequest $request;
+  private IUserManager $userManager;
 	private IClientService $clientService;
 	private IConfig $config;
-	private LoggerInterface $logger;
+  private LoggerInterface $logger;
+  private Manager $twoFactorManager;
 
 	public function __construct(
 		GlobalSiteSelector $gss,
 		ICrypto $crypto,
 		Lookup $lookup,
-		IRequest $request,
+    IRequest $request,
+    IUserManager $userManager,
 		IClientService $clientService,
 		IConfig $config,
-		LoggerInterface $logger
+    LoggerInterface $logger,
+    Manager $twoFactorManager
 	) {
 		$this->gss = $gss;
 		$this->crypto = $crypto;
 		$this->lookup = $lookup;
-		$this->request = $request;
+    $this->request = $request;
+    $this->userManager = $userManager;
 		$this->clientService = $clientService;
 		$this->config = $config;
-		$this->logger = $logger;
+    $this->logger = $logger;
+    $this->twoFactorManager = $twoFactorManager;
 	}
 
 
@@ -169,7 +177,14 @@
 			}
 		}
 
-		if (!empty($location)) {
+    if (!empty($location)) {
+      if (filter_var($this->config->getAppValue('globalsiteselector', 'enable_mfa'), FILTER_VALIDATE_BOOLEAN)) {
+        $this->logger->debug(
+          'handleLoginRequest: preparing 2fa for: ' . $uid . ', before redirecting'
+        );
+        $user = $this->userManager->get($uid);
+        $this->twoFactorManager->prepareTwoFactorLogin($user, false);
+      }
 			$this->logger->debug(
 				'handleLoginRequest: redirecting user: ' . $uid . ' to ' . $this->normalizeLocation($location)
 			);
