diff --git a/lib/Master.php b/lib/Master.php
index 1c410f8..c5a1d0f 100644
--- a/lib/Master.php
+++ b/lib/Master.php
@@ -30,11 +30,13 @@ use Firebase\JWT\Key;
 use OCA\GlobalSiteSelector\AppInfo\Application;
 use OCA\GlobalSiteSelector\UserDiscoveryModules\IUserDiscoveryModule;
 use OCP\Authentication\IApacheBackend;
+use OC\Authentication\TwoFactorAuth\Manager;
 use OCP\HintException;
 use OCP\Http\Client\IClientService;
 use OCP\IConfig;
 use OCP\IRequest;
 use OCP\ISession;
+use OCP\IUserManager;
 use OCP\Security\ICrypto;
 use OCP\Server;
 use Psr\Container\ContainerExceptionInterface;
@@ -54,28 +56,33 @@ class Master {
 	private ICrypto $crypto;
 	private Lookup $lookup;
 	private IRequest $request;
+  private IUserManager $userManager;
 	private IClientService $clientService;
 	private IConfig $config;
 	private LoggerInterface $logger;
-
+  private Manager $twoFactorManager;
 	public function __construct(
 		ISession $session,
 		GlobalSiteSelector $gss,
 		ICrypto $crypto,
 		Lookup $lookup,
 		IRequest $request,
+    IUserManager $userManager,
 		IClientService $clientService,
 		IConfig $config,
-		LoggerInterface $logger
+    LoggerInterface $logger,
+    Manager $twoFactorManager
 	) {
 		$this->session = $session;
 		$this->gss = $gss;
 		$this->crypto = $crypto;
 		$this->lookup = $lookup;
 		$this->request = $request;
+		$this->userManager = $userManager;
 		$this->clientService = $clientService;
 		$this->config = $config;
-		$this->logger = $logger;
+    $this->logger = $logger;
+    $this->twoFactorManager = $twoFactorManager;
 	}
 
 
@@ -175,8 +182,14 @@ class Master {
 				);
 			}
 		}
-
 		if (!empty($location)) {
+      if (filter_var($this->config->getAppValue('globalsiteselector', 'enable_mfa'), FILTER_VALIDATE_BOOLEAN)) {
+        $this->logger->debug(
+          'handleLoginRequest: preparing 2fa for: ' . $uid . ', before redirecting'
+        );
+        $user = $this->userManager->get($uid);
+        $this->twoFactorManager->prepareTwoFactorLogin($user, false);
+      }
 			$this->logger->debug(
 				'handleLoginRequest: redirecting user: ' . $uid . ' to ' . $this->normalizeLocation($location)
 			);
