diff --git a/lib/private/Files/Storage/Common.php b/lib/private/Files/Storage/Common.php
index 3473eebc4a0e..1f849f8e29c4 100644
--- a/lib/private/Files/Storage/Common.php
+++ b/lib/private/Files/Storage/Common.php
@@ -761,6 +761,7 @@ public function acquireLock($path, $type, ILockingProvider $provider) {
 		try {
 			$provider->acquireLock('files/' . md5($this->getId() . '::' . trim($path, '/')), $type, $this->getId() . '::' . $path);
 		} catch (LockedException $e) {
+			$e = new LockedException($e->getPath(), $e, $e->getExistingLock(), $path);
 			if ($logger) {
 				$logger->info($e->getMessage(), ['exception' => $e]);
 			}
@@ -793,6 +794,7 @@ public function releaseLock($path, $type, ILockingProvider $provider) {
 		try {
 			$provider->releaseLock('files/' . md5($this->getId() . '::' . trim($path, '/')), $type);
 		} catch (LockedException $e) {
+			$e = new LockedException($e->getPath(), $e, $e->getExistingLock(), $path);
 			if ($logger) {
 				$logger->info($e->getMessage(), ['exception' => $e]);
 			}
@@ -825,6 +827,7 @@ public function changeLock($path, $type, ILockingProvider $provider) {
 		try {
 			$provider->changeLock('files/' . md5($this->getId() . '::' . trim($path, '/')), $type);
 		} catch (LockedException $e) {
+			$e = new LockedException($e->getPath(), $e, $e->getExistingLock(), $path);
 			if ($logger) {
 				$logger->info($e->getMessage(), ['exception' => $e]);
 			}
