diff --git a/apps/dav/lib/Connector/Sabre/File.php b/apps/dav/lib/Connector/Sabre/File.php
index 94632b265db..669b3379f3a 100644
--- a/apps/dav/lib/Connector/Sabre/File.php
+++ b/apps/dav/lib/Connector/Sabre/File.php
@@ -199,8 +199,14 @@ class File extends Node implements IFile {
 					// the entire operation will be safe
 
 					try {
+						\OCP\Log\logger('webdav')->warning('Re-acquiring log for large upload without part file', [
+							'exception' => $e,
+						]);
 						$this->acquireLock(ILockingProvider::LOCK_EXCLUSIVE);
 					} catch (LockedException $ex) {
+						\OCP\Log\logger('webdav')->error('Failed to re-acquiring log for large upload without part file', [
+							'exception' => $e,
+						]);
 						throw new FileLocked($e->getMessage(), $e->getCode(), $e);
 					}
 				}
@@ -336,8 +342,14 @@ class File extends Node implements IFile {
 					// the entire operation will be safe
 
 					try {
+						\OCP\Log\logger('webdav')->warning('Re-acquiring log for large upload with part file', [
+							'exception' => $e,
+						]);
 						$this->acquireLock(ILockingProvider::LOCK_EXCLUSIVE);
 					} catch (LockedException $ex) {
+						\OCP\Log\logger('webdav')->error('Failed to re-acquiring log for large upload with part file', [
+							'exception' => $e,
+						]);
 						if ($needsPartFile) {
 							$partStorage->unlink($internalPartPath);
 						}
