diff -bur old/lib/private/Files/ObjectStore/S3ConnectionTrait.php new/lib/private/Files/ObjectStore/S3ConnectionTrait.php
--- old/lib/private/Files/ObjectStore/S3ConnectionTrait.php	2024-01-26 10:31:00.031243802 +0100
+++ new/lib/private/Files/ObjectStore/S3ConnectionTrait.php	2024-01-26 10:31:40.877975638 +0100
@@ -74,7 +74,7 @@
 	/** @var int */
 	private $copySizeLimit;
 
-	private bool $useMultipartCopy = true;
+	private bool $useMultipartCopy = false;
 
 	protected $test;
 
@@ -93,7 +93,8 @@
 		$this->uploadPartSize = $params['uploadPartSize'] ?? 524288000;
 		$this->putSizeLimit = $params['putSizeLimit'] ?? 104857600;
 		$this->copySizeLimit = $params['copySizeLimit'] ?? 5242880000;
-		$this->useMultipartCopy = (bool)($params['useMultipartCopy'] ?? true);
+		// $this->useMultipartCopy = (bool)($params['useMultipartCopy'] ?? true);
+                $this->useMultipartCopy = false;
 		$params['region'] = empty($params['region']) ? 'eu-west-1' : $params['region'];
 		$params['hostname'] = empty($params['hostname']) ? 's3.' . $params['region'] . '.amazonaws.com' : $params['hostname'];
 		if (!isset($params['port']) || $params['port'] === '') {
