diff --git a/apps/dav/lib/CardDAV/SystemAddressbook.php b/apps/dav/lib/CardDAV/SystemAddressbook.php
index 502e353acb3..1c78e03162e 100644
--- a/apps/dav/lib/CardDAV/SystemAddressbook.php
+++ b/apps/dav/lib/CardDAV/SystemAddressbook.php
@@ -30,6 +30,7 @@
 use OCP\IConfig;
 use OCP\IL10N;
 use Sabre\CardDAV\Backend\BackendInterface;
+use function array_filter;
 
 class SystemAddressbook extends AddressBook {
 	/** @var IConfig */
@@ -50,4 +51,13 @@ public function getChildren() {
 
 		return parent::getChildren();
 	}
+
+	public function getACL() {
+		return array_filter(parent::getACL(), function($acl) {
+			if (in_array($acl['privilege'], ['{DAV:}write', '{DAV:}all'], true)) {
+				return false;
+			}
+			return true;
+		});
+	}
 }
diff --git a/apps/files_external/lib/Lib/Auth/Password/SessionCredentials.php b/apps/files_external/lib/Lib/Auth/Password/SessionCredentials.php
index dbe5a2fdc20..228366db204 100644
--- a/apps/files_external/lib/Lib/Auth/Password/SessionCredentials.php
+++ b/apps/files_external/lib/Lib/Auth/Password/SessionCredentials.php
@@ -30,6 +30,7 @@
 use OCP\Authentication\Exceptions\CredentialsUnavailableException;
 use OCP\Authentication\LoginCredentials\IStore as CredentialsStore;
 use OCP\Files\Storage;
+use OCP\Files\StorageAuthException;
 use OCP\IL10N;
 use OCP\IUser;
 
@@ -57,6 +58,10 @@ public function manipulateStorageConfig(StorageConfig &$storage, IUser $user = n
 			throw new InsufficientDataForMeaningfulAnswerException('No session credentials saved');
 		}
 
+		if ($credentials->getUID() !== $user->getUID()) {
+			throw new StorageAuthException('Session credentials for storage owner not available');
+		}
+
 		$storage->setBackendOption('user', $credentials->getLoginName());
 		$storage->setBackendOption('password', $credentials->getPassword());
 	}
