diff --git a/lib/AppInfo/Application.php b/lib/AppInfo/Application.php
index dbe3402..7d53c22 100644
--- a/lib/AppInfo/Application.php
+++ b/lib/AppInfo/Application.php
@@ -31,7 +31,7 @@ use OCP\AppFramework\App;
 use OCP\AppFramework\Bootstrap\IBootContext;
 use OCP\AppFramework\Bootstrap\IBootstrap;
 use OCP\AppFramework\Bootstrap\IRegistrationContext;
-use OCP\User\Events\UserLoggedInEvent;
+use OCA\GlobalSiteSelector\Events\AfterLoginOnSlaveEvent;
 use Throwable;
 
 
@@ -53,7 +53,7 @@ class Application extends App implements IBootstrap
    */
   public function register(IRegistrationContext $context): void
   {
-    $context->registerEventListener(UserLoggedInEvent::class, UserLoggingIn::class);
+    $context->registerEventListener(AfterLoginOnSlaveEvent::class, UserLoggingIn::class);
   }
   /**
    * @param IBootContext $context
diff --git a/lib/Listeners/UserLoggingIn.php b/lib/Listeners/UserLoggingIn.php
index 629d50e..74e0bf2 100644
--- a/lib/Listeners/UserLoggingIn.php
+++ b/lib/Listeners/UserLoggingIn.php
@@ -30,7 +30,7 @@ use OCP\EventDispatcher\Event;
 use OCP\EventDispatcher\IEventListener;
 use OCP\IUser;
 use OCP\ISession;
-use OCP\User\Events\UserLoggedInEvent;
+use OCA\GlobalSiteSelector\Events\AfterLoginOnSlaveEvent;
 use Psr\Log\LoggerInterface;
 
 
@@ -53,7 +53,7 @@ class UserLoggingIn implements IEventListener
    */
   public function handle(Event $event): void
   {
-    if (!$event instanceof UserLoggedInEvent) {
+    if (!$event instanceof AfterLoginOnSlaveEvent) {
       return;
     }
     /**
